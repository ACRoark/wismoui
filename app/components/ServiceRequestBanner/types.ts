export interface IBannerNotification {
  hasError: boolean;
  message: IMessage;
}

export interface IMessage {
  id: string;
  values: string[] | null;
}
