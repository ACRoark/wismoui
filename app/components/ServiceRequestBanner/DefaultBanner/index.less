@import '../../../styles/themes/index.less';

@service-request-banner-cls-prefix: ~'@{dte-prefix}-default-banner';
@mobile: ~'only screen and (max-width: 365px)';
@desktop: ~'only screen and (min-width: 365px)';

.@{service-request-banner-cls-prefix} {
  background-color: @dte-gray-200;
  margin-bottom: 48px;

  > h4 {
    display: inline-block;
    font-weight: @dte-wismo-font-weight-semi-bold;
    line-height: 1.5;
    margin-block-end: 0px;
    margin-block-start: 0px;
    overflow-wrap: normal;
    padding: 20px 30px;
  }
}
