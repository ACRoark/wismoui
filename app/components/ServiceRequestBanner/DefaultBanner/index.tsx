import React, { ReactElement } from 'react';
import { FormattedMessage } from 'react-intl';

import DateStamp from 'components/DateStamp';
import messages from 'components/ServiceRequestBanner/messages';

import { IMessage } from '../types';
import './index.less';

interface IDefaultBannerProps {
  message: IMessage;
}

const getMessage = (id: string, values: string[] | null): ReactElement => {
  if (values) {
    if (values.length > 1) {
      return (
        <FormattedMessage
          {...messages[id]}
          values={{
            date1: <DateStamp value={values[0]} />,
            date2: <DateStamp value={values[1]} />,
          }}
        />
      );
    }

    return <FormattedMessage {...messages[id]} values={{ date: <DateStamp value={values[0]} /> }} />;
  }

  return <FormattedMessage {...messages[id]} values={{ date: null }} />;
};

const DefaultBanner: React.FC<IDefaultBannerProps> = (props: IDefaultBannerProps): React.ReactElement => {
  const {
    message: { id, values },
  } = props;

  const message = getMessage(id, values);

  return (
    <div className="dte-wismo-default-banner">
      <h4>{message}</h4>
    </div>
  );
};

export default DefaultBanner;
