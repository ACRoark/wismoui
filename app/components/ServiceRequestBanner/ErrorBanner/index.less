@import '../../../styles/themes/index.less';

@service-request-error-banner-cls-prefix: ~'@{dte-prefix}-error-banner';

.@{service-request-error-banner-cls-prefix} {
  background-color: @dte-background-error;
  display: flex;
  flex-direction: row;
  margin-bottom: 48px;
  padding: 20px;

  > div > h4 {
    color: @dte-red !important;
    display: inline;
    font-weight: @dte-wismo-font-weight-semi-bold;
    line-height: 1.5;
    margin-block-end: 0;
    margin-block-start: 0;
    margin-right: 20px;
    overflow-wrap: normal;
  }

  .content {
    margin-left: 20px;

    @media @mobile {
      align-items: flex-start;
      display: flex;
      flex-direction: column;
    }
  }

  .error-icon {
    color: @dte-red;
    display: inline-block;
    margin: 50% 0;
    vertical-align: text-top;
  }

  .view-details-link {
    font-size: @dte-wismo-font-size-0;
    font-weight: @dte-wismo-font-weight-semi-bold;
    padding: unset !important;
  }
}
