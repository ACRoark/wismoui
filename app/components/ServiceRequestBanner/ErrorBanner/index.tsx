import React from 'react';
import { FormattedMessage } from 'react-intl';

import ExclamationPoint from 'images/ExclamationPoint';

import messages from '../messages';
import { IMessage } from '../types';
import './index.less';

interface IErrorBannerProps {
  message: IMessage;
}

const ErrorBanner: React.FC<IErrorBannerProps> = (props: IErrorBannerProps): React.ReactElement => {
  const {
    message: { id },
  } = props;

  return (
    <div className="dte-wismo-error-banner">
      <div>
        <ExclamationPoint className="error-icon" title="error" />
      </div>
      <div className="content">
        <h4>
          <FormattedMessage {...messages[id]} values={{ date: null }} />
        </h4>
      </div>
    </div>
  );
};

export default ErrorBanner;
