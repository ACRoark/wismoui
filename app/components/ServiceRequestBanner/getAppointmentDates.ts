import { IAppointmentList } from 'types';

const getAppointmentDates = (appointmentList: IAppointmentList[]): string[] => {
  const dates: string[] = [];

  appointmentList.forEach((appointment: IAppointmentList): void => {
    if (!appointment.completed) {
      dates.push(appointment.date);
    }
  });

  return dates;
};

export default getAppointmentDates;
