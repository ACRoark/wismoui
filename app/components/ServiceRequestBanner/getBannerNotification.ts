import * as _ from 'lodash';

import createAppointmentList from 'components/Stepper/createAppointmentList';
import { IAppointmentList, IOrderRequest, IServiceOrderEvent } from 'types';
import filterVerificationEvents from 'utils/filterVerificationEvents';

import getAppointmentDates from './getAppointmentDates';

import { IBannerNotification } from './types';

const getBannerNotification = (serviceRequest: IOrderRequest): IBannerNotification => {
  const { verificationEvents } = serviceRequest;

  const appointmentList: IAppointmentList[] = createAppointmentList(serviceRequest);
  const dates = getAppointmentDates(appointmentList);
  const hasMultipleDates = dates.length > 1;

  const verificationCodes = ['AD', 'AN', 'CL', 'DR', 'ES', 'FV', 'FW', 'LD', 'PR', 'RE', 'RF', 'SC', 'WP'];
  const filteredVerificationEvents = filterVerificationEvents(verificationEvents, verificationCodes);

  if (filteredVerificationEvents.length) {
    const id = _.last(filteredVerificationEvents[0].verificationEventStatusUpdates).verificationCode;

    return {
      hasError: true,
      message: {
        id,
        values: null,
      },
    };
  }

  const allAppointments: IServiceOrderEvent[] = appointmentList.reduce(
    (appointments: IServiceOrderEvent[], appointmentsByDate: IAppointmentList): IServiceOrderEvent[] => {
      if (!appointmentsByDate.completed) {
        return appointments.concat(appointmentsByDate.appointments);
      }
      return appointments;
    },
    [],
  );

  const delayedErrorAppointment = allAppointments.find((appointment: IServiceOrderEvent): boolean =>
    ['DELAYED'].includes(`${appointment.latestStatus.serviceOrderEventStatus}`),
  );

  const errorAppointment = allAppointments.find((appointment: IServiceOrderEvent): boolean =>
    ['CALL_DTE', 'CANCELED_CGI', 'CANCELED_INC', 'PARTIAL_COMPLETION', 'RESCHEDULED'].includes(
      `${appointment.latestStatus.serviceOrderEventStatus}`,
    ),
  );

  if (delayedErrorAppointment || errorAppointment) {
    const id = `${
      delayedErrorAppointment?.latestStatus.serviceOrderEventStatus ||
      errorAppointment?.latestStatus.serviceOrderEventStatus
    }`;

    return {
      hasError: true,
      message: {
        id,
        values: null,
      },
    };
  }

  return {
    hasError: false,
    message: {
      id: hasMultipleDates ? `${serviceRequest.orderRequestType}_multiple_dates` : serviceRequest.orderRequestType,
      values: dates.length === 0 ? null : dates,
    },
  };
};

export default getBannerNotification;
