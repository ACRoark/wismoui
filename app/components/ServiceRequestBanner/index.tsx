import React from 'react';

import { IOrderRequest } from 'types';

import DefaultBanner from './DefaultBanner';
import ErrorBanner from './ErrorBanner';
import getBannerNotification from './getBannerNotification';
import { IBannerNotification } from './types';

interface IServiceRequestBannerProps {
  serviceRequest: IOrderRequest;
}

const ServiceRequestBanner: React.FC<IServiceRequestBannerProps> = (
  props: IServiceRequestBannerProps,
): React.ReactElement => {
  const { serviceRequest } = props;

  const bannerNotification: IBannerNotification = getBannerNotification(serviceRequest);

  if (bannerNotification.hasError) {
    return <ErrorBanner message={bannerNotification.message} />;
  }

  return <DefaultBanner message={bannerNotification.message} />;
};

export default ServiceRequestBanner;
