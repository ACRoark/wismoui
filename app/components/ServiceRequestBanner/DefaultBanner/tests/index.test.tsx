import React from 'react';

import { SUPPORTED_LOCALES } from 'locales';

import createFakeOrder from 'components/createFakeOrder';
import getBannerNotification from 'components/ServiceRequestBanner/getBannerNotification';
import {  IMessage } from 'components/ServiceRequestBanner/types';
import { createSnapshotWithIntl, renderWithIntl } from 'testing/utils';
import { IOrderRequest } from 'types';

import DefaultBanner from '../index';

const fakeStartService = createFakeOrder('MI1122223333');
const fakeStopService = createFakeOrder('MO2233334444');

const expectComponentToMatchSnapshot = (locale: string, message: IMessage): void => {
  const tree = createSnapshotWithIntl(<DefaultBanner message={message} />, locale);

  expect(tree).toMatchSnapshot();
};

describe('<DefaultBanner />', (): void => {
  const languages = Object.keys(SUPPORTED_LOCALES);

  languages.forEach((language: string): void => {
    it(`should not log errors in console when rendered with a Start Service request and the current language is ${language}`, (): void => {
      const spy = jest.spyOn(global.console, 'error');

      const { message } = getBannerNotification(fakeStartService.orderRequests[0]);

      renderWithIntl(<DefaultBanner message={message} />, language);

      expect(spy).not.toHaveBeenCalled();
    });

    it(`should render correctly when given a Start Service request and the current language is ${language}`, (): void => {
      const { message } = getBannerNotification(fakeStartService.orderRequests[0]);

      expectComponentToMatchSnapshot(language, message);
    });

    it(`should render correctly when given a Stop Service request and the current language is ${language}`, (): void => {
      const { message } = getBannerNotification(fakeStopService.orderRequests[0]);

      expectComponentToMatchSnapshot(language, message);
    });

    it(`should render correctly when given a Start Service request with multiple appointment dates and the current language is ${language}`, (): void => {
      const serviceRequest: IOrderRequest = {
        ...fakeStartService.orderRequests[0],
        serviceOrderEvents: [
          {
            appointment: {
              date: '20220522',
              slotType: 'AM',
            },
            latestStatus: {
              createdAtDate: '20200522',
              createdAtTime: '223443',
              serviceOrderEventStatus: 'ON_SCHEDULE',
            },
            product: 'EFR_D1',
            serviceOrderCategory: null,
            serviceOrderId: 'abc123',
          },
          {
            appointment: {
              date: '20220523',
              slotType: 'PM',
            },
            latestStatus: {
              createdAtDate: '20200522',
              createdAtTime: '223443',
              serviceOrderEventStatus: 'ON_SCHEDULE',
            },
            product: 'GFR_AS_H',
            serviceOrderCategory: null,
            serviceOrderId: 'abc123',
          },
        ],
      };

      const { message } = getBannerNotification(serviceRequest);

      expectComponentToMatchSnapshot(language, message);
    });
  });
});
