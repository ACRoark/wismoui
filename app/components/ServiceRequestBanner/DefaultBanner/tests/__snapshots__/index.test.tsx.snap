// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DefaultBanner /> should render correctly when given a Start Service request and the current language is en 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Start Service 
    Sunday
    , 
    May
     
    22
  </h4>
</div>
`;

exports[`<DefaultBanner /> should render correctly when given a Start Service request and the current language is es 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Comience el serviciom 
    Domingo
    , 
    Mayo
     
    22
  </h4>
</div>
`;

exports[`<DefaultBanner /> should render correctly when given a Start Service request with multiple appointment dates and the current language is en 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Start Service 
    Sunday
    , 
    May
     
    22
     & 
    Monday
    , 
    May
     
    23
  </h4>
</div>
`;

exports[`<DefaultBanner /> should render correctly when given a Start Service request with multiple appointment dates and the current language is es 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Comience el serviciom 
    Domingo
    , 
    Mayo
     
    22
     & 
    Lunes
    , 
    Mayo
     
    23
  </h4>
</div>
`;

exports[`<DefaultBanner /> should render correctly when given a Stop Service request and the current language is en 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Stop Service 
    Sunday
    , 
    May
     
    22
  </h4>
</div>
`;

exports[`<DefaultBanner /> should render correctly when given a Stop Service request and the current language is es 1`] = `
<div
  className="dte-wismo-default-banner"
>
  <h4>
    Parada de servicio 
    Domingo
    , 
    Mayo
     
    22
  </h4>
</div>
`;
